[ENH](WIP) new ufunc implementations for object arrays

ahaldane opened the pull request:
This PR overhauls the ufunc implementations for object type, to follow the rules I will paste in the comment below.

At this point I am pasting it to see if it there is interest so it is worth effort to finish unit tests, and to discuss what the ufunc implementations should be. I still need to do tests, docs, and bug-hunting.
### motivation

Currently many ufuncs do not work on object arrays (eg all transcendental functions), and some that do work can be improved (see recent discussion of `np.sign`). With this PR all (or most...) missing ufuncs are implemented using the `math` and `cmath` modules as a fallback.
### request for comments

I am sure I have missed funny edge cases in the ufunc implementations, so I would love more eyes looking at them. The current implementations are in `_objectmath.py`. 

Some issues in particular: I am not sure how to implement `np.nextafter`, `np.spacing` or `np.cbrt` in pure python. Also, I changed the way the `logical_` ufuncs work: In current numpy, `logical_and` (for example) behave like `x and y` (Ie, it will return either x or y), while I changed it to `True if (x and y) else False` (so it returns a boolean). The new way behaves more like the ufuncs behave for float/int arrays.

Also, while we are updating the ufunc implementation, maybe this is the time for back-compat breaks. For example, should the fallback to (eg) `sqrt` be to call the `elem.sqrt` method, or might we rename it to `elem.npy_sqrt` to avoid name conflicts?

ahaldane committed:
Redo Ufuncs for Object Arrays

ahaldane commented:
Here is what this PR does. (This can go into Docs later)

ufuncs and object arrays
========================

Ufuncs operate specially on objects arrays, since the objects contained in the array may not be numpy types. To evaluate a ufunc numpy tries a series of strategies in order for each element of the array (for unary ufuncs) or for each pair of elements from the arrays (for binary ufuncs): 

1. If there is a Python ["Special method"](https://docs.python.org/2/reference/datamodel.html#special-method-names) which corresponds to the ufunc numpy uses it to evaluate the ufunc. This step only applies to the ufuncs `add`, `subtract`, `multiply`, `divide`, `true_divide`, `floor_divide`, `remainder`, `negative`, `[positive]`, `power`, `mod`, `absolute`, `bitwise_and`, `bitwise_or`, `bitwise_xor`, `invert`, `left_shift`, `right_shift`, `greater`, `greater_equal`, `less`, `less_equal`, `not_equal`, `equal`.

2. If any of the arguments is a numpy scalar or ndarray then numpy applies the ufunc directly, eg `np.sqrt(elem)`.

3. If the first argument has a method with the same name as the ufunc then that method will be called, eg `elem.sqrt()`. Binary ufuncs such as `logaddexp(x,y)` will call `x.logaddexp(y)`.

4. Numpy will then fall back to a Python implementation of the ufunc which may use the `math` and `cmath` modules, eg `math.sqrt(elem)`. This includes all ufuncs not evaluated in step 1. Details of the Python implementations are given in [the file `_obectmath.py`].

If no step succeeds, a `TypeError` error is raised. [This is not implemented yet, not sure if we want this]

This provides a rough way of creating new numeric types compatible with numpy, as you can define a class which implements all ufuncs missing from step 1 as methods, and then create an object array containing elements of your type. However, note that creating a user-defined type (see "User-Defined Types") is often preferrable. User defined types will be much faster and you will have control over casting and coercion.

njsmith commented:
Can you post this to the list, so more eyes will see it?
On Sep 15, 2015 11:49 AM, "ahaldane" <notifications@github.com> wrote:

> Here is what this PR does. (This can go into Docs later)
> ufuncs and object arrays
>
> Ufuncs operate specially on objects arrays, since the objects contained in
> the array may not be numpy types. To evaluate a ufunc numpy tries a series
> of strategies in order for each element of the array (for unary ufuncs) or
> for each pair of elements from the arrays (for binary ufuncs):
>
>    1.
>
>    If there is a Python "Special method"
>    <https://docs.python.org/2/reference/datamodel.html#special-method-names>
>    which corresponds to the ufunc numpy uses it to evaluate the ufunc. This
>    step only applies to the ufuncs add, subtract, multiply, divide,
>    true_divide, floor_divide, remainder, negative, [positive], power, mod,
>    absolute, bitwise_and, bitwise_or, bitwise_xor, invert, left_shift,
>    right_shift, greater, greater_equal, less, less_equal, not_equal, equal
>    .
>    2.
>
>    If any of the arguments is a numpy scalar or ndarray then numpy
>    applies the ufunc directly, eg np.sqrt(elem).
>    3.
>
>    If the first argument has a method with the same name as the ufunc
>    then that method will be called, eg elem.sqrt(). Binary ufuncs such as
>    logaddexp(x,y) will call x.logaddexp(y).
>    4.
>
>    Numpy will then fall back to a Python implementation of the ufunc
>    which may use the math and cmath modules, eg math.sqrt(elem). This
>    includes all ufuncs not evaluated in step 1. Details of the Python
>    implementations are given in [the file _obectmath.py].
>
> If no step succeeds, a TypeError error is raised. [This is not
> implemented yet, not sure if we want this]
>
> This provides a rough way of creating new numeric types compatible with
> numpy, as you can define a class which implements all ufuncs missing from
> step 1 as methods, and then create an object array containing elements of
> your type. However, note that creating a user-defined type (see
> "User-Defined Types") is often preferrable. User defined types will be much
> faster and you will have control over casting and coercion.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6320#issuecomment-140496431>.
>

ahaldane commented:
Sure.

Also:

 * I don't know why some python3 builds are failing. It doesn't like the macro expansion, but I did it exactly the same as the other ufuncs in that file... I'll have to figure it out later.
 * Part of the motivation for the 4 steps above is so that all ufuncs are customizable: Step 1 methods by defining the special method, and everything else through step 3. In current numpy there are many ufuncs handled in step 1 (eg `sign`, `maximum`) which the user cannot override.

ewmoore commented on numpy/core/_objectmath.py:
```diff
@@ -0,0 +1,188 @@
+
+def signbit(x):
+    if isinstance(x, complex):
+        return cmath.copysign(1, x) < 0
```
`cmath.copysign` isn't a thing.

ewmoore commented:
#368 has some earlier discussion of this idea.

ahaldane committed:
part 2

seberg commented:
I hate to do this, but my question is: what about for example decimals?

ahaldane commented:
That's exactly the kind of thing I was worried about missing. I hadn't thought of Decimals.

It looks like they are already mostly handled.
 
 * add/multiply are covered by special methods (step 1). 
 * The `logical_` funcs, `log`, `exp`, `sqrt` and more are covered in step 3 since they are Decimal methods. 
 * The math module accepts them (converting to float), covering many remaining ufuncs (sort of).

Some ufuncs could be improved, eg using the `is_nan`, `is_inf`, `copy_sign` methods of Decimal, or using the `exp` method to implement `expm1`.

But, should we even try to support Decimal? It doesn't play well with the other numeric types. For example, `Decimal(1) + 1.0` fails. Perhaps if someone wants to use Decimal object arrays, they should subclass/wrap Decimal in a way to make numpy happy.

ahaldane committed:
Part 3

seberg commented:
The thing I am worried about is not about them not working (I am fine with that really). What I have a bit of a bad feeling about is silently casting decimals to float, which to me feels wrong (it is an unsafe cast). Ideally, you we would probably make it easy to create a full dtype for it, but the "object" dtype should play well with them in any case.

In other words, I am not quite comfortable with calling into `math` or `cmath` unless we got the vanilla python `float`/`complex` type. `Decimal`s have `exp`, but you might end up calling `math.exp` on them, because they are not a complex type (you should use `numbers.Complex` there in any case, but maybe the explicit is good, just also do the `float` explicitly for `math`?).

ahaldane commented:
The same problem arises for people using the mpmath types. My original thought was that if an object implemented `__float__` it was probably happy for us to convert to float if needed. 

But now I am thinking you are right, it would be confusing. Maybe we should have something like

    def _notimplemented_msg(fname, etype):
        return ("fallback for ufunc '{}' is not implemented "
                "for objects of type {}").format(fname, etype)

    def exp(x):
        if isinstance(x, complex):
            return cmath.exp(x)
        if isinstance(x, (int, long, float)):
            return math.exp(x)
        raise TypeError(_notimplemented_msg('exp', type(x)))


(Note that `isinstance(True, int)` is True, which I think is what we want)

ahaldane commented:
Also, what do you think of default implementation in terms of other ufuncs? Eg, what if we have

    def expm1(x):
        if isinstance(x, (int, long, float)):
            return math.expm1(x)
        return _ufunc_exp(x) - 1

where `_ufunc_exp` is essentially `np.exp`, so it goes back through the ufunc object dispatch code. This way, `np.exp(Decimal(1))` will effectively call `Decimal(1).exp() - 1`. 

On the plus side, it means the user doesn't need to implement `expm1` if `exp` is implemented. On the other hand, it silently uses a less precise calculation.

njsmith commented:
It sounds like a lot of the use cases that this is aiming at would be
better handled by a new dtype system that lets us have an actual Decimal or
mpmath dtype, instead of trying to guess what to do with arbitrary objects
in a one-size-fits-all way. (Which doesn't necessarily mean we should block
bandaid solutions now -- bandaids are good first aid! -- but something to
keep in mind while thinking about this.)
On Sep 16, 2015 11:43 AM, "ahaldane" <notifications@github.com> wrote:

> Also, what do you think of default implementation in terms of other
> ufuncs? Eg, what if we have
>
> def expm1(x):
>     if isinstance(x, (int, long, float)):
>         return math.expm1(x)
>     return _ufunc_exp(x) - 1
>
> where _ufunc_exp is essentially np.exp, so it goes back through the ufunc
> object dispatch code. This way, np.exp(Decimal(1)) will effectively call Decimal(1).exp()
> - 1.
>
> On the plus side, it means the user doesn't need to implement expm1 if exp
> is implemented. On the other hand, it silently uses a less precise
> calculation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6320#issuecomment-140838745>.
>

ahaldane commented:
Updated. Everyone's comments make me think a lot of ufuncs should just fail for types that aren't builtin (bool/int, long, float or complex), so now they raise `TypeErrors`. It's not 100% consistent though - `isfinite` does not need to be implemented as long as `isinf` and `isnan` are. `log2` is implemented in terms of log. `fmax`, `fmin` are similar. I'm still on the fence about these. 

Ufuncs which exist to increase precision (`expm1`, `log1p`, `logaddexp`) will raise TypeErrors for non-builtin types. I don't want to get in the buisness of guessing a numerically accurate implementation when I don't know what the input type is.

seberg commented:
@njsmith, while I do think that a lot of this would fit much better with some better dtype system, we should have reasonable stuff in place for plain `object` arrays (the question is what that is ;)).
I can live with specializing python float/long/int as long as we do not cast things like `decimals`. I think this has to be run by the list (which I guess it already was, but the list is hanging)....

I think it might be better if we do not use isinstance, but actually `type(blah) in (int, long, float)`. But I am willing to concede here if everyone thinks it is just so much more practical. Another example of these kind of problems:
```
In [3]: m = np.finfo(np.float128).max
In [4]: import math
In [5]: math.isinf(m)
Out[5]: True
```

So my question is: unless it is completely obvious that one thing is right (sign seems fine to me for example, though then again, strings ^^, so maybe a "number" check....), how about we always check the type exactly. That feels pretty future proof to me in terms of not breaking down at some point and does  probably solve almost all real live cases?!

ahaldane commented:
I think you're right about not using isinstance. That will be a problem for mpmath types too, eg `math.isinf(mpmath.mpf(10)**800)` says `True`. 

Also `sign` currently doesn't work for complex numbers. Actually `np.sign` is a little weird in that case already: 
    
    >>> np.sign(0)
    0
    >> np.sign(0+1j)
    (1+0j)

argriffing commented:
> Actually np.sign is a little weird in that case already

see https://github.com/numpy/numpy/issues/1848

ahaldane commented:
Ah! But now that I get down to implementing it, is this what we want?

    >>> np.sign(np.inf*1j)
    (nan+0j)

I would expect +1 based on wikipedia's definition since `np.sign(np.inf)` gives +1.

pv commented:
@ahaldane: `np.inf*1j -> (nan+infj)`. In c99 there's only one complex infinity (it has multiple representations). For `z/sqrt(z**2)`, infinity is a similarly singular point as 0, so sign is probably not defined there. Unlike 0, it probably cannot be defined as 0 either...

pv commented:
But it probably would also be fine to define it in terms of whatever limiting procedure, which might be less surprising for some cases.

ahaldane commented:
@pv, that makes sense. But also, in python and numpy I can specify the type of infinity:

    >>> np.sign(complex(0, -np.inf))
    (-1+0j)

But this is not very accurate then:

    >>> repr(np.complex64(complex(0, -np.inf)))
    '-inf*j'

ahaldane committed:
Part 4

ahaldane commented:
All right, I updated all the ufunc implementations based on recent discussion, fixed small bugs in the C code, and also added a rough unit test.

The unit test checks that the ufunc works similarly for numpy types and python builtin types. 

However, it turns out a bunch of my ufunc implementations failed the test. I fixed some of them, but others I've just disabled for now in the unit test. See the comments in the unit test.

ahaldane committed:
Part 5

ahaldane commented:
If we are only going to handle `bool, int, long, float, complex`, it is starting to make sense to me that we should just do this for all ufuncs:

    def ufunc(x):
        if type(x) in (bool, int, long, float, complex):
            return np.ufunc(x).item()
        raise TypeError

njsmith commented:
Interestingly, that definition also makes sense for ndarrays and eventually
anything with __numpy_ufunc__.
On Sep 17, 2015 3:34 PM, "ahaldane" <notifications@github.com> wrote:

> If we are only going to handle bool, int, long, float, complex, it is
> starting to make sense to me that we should just do this for all ufuncs:
>
> def ufunc(x):
>     if type(x) in (bool, int, long, float, complex):
>         return np.ufunc(x).item()
>     raise TypeError
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6320#issuecomment-141256231>.
>

seberg commented:
Actually, I think it would likely make sense to define "object" arrays in the sense of "try to call the type specific ufunc -- if we know it -- on each element" + some fallback (unless they map to a python operator maybe).
Just to annoy you, we might want to support numpy dtypes as well. Whether or not they may be cast, I am not sure ;). The ufuncs have slightly different behaviours sometimes I think, but yes, the only objections to it I can think of, is that is possibly a bit slow.

njsmith commented:
For most objects that support both + and np.add they should do the same
thing. (Whether this should be true of *all* objects was one of the key
questions unresolved from #5844.) There are some objects that only support
+, and under two of the three main approaches in #5844, anything that
supports np.add should generally do the same thing for +. so it might make
sense to just use binops unconditionally for the ufuncs that have
corresponding binops, and ufuncs otherwise.
.
This does get more confusing though if we go with "option 3" there and
allow + and np.add to do arbitrarily different things -- in this case it'd
be sort of weird to call add on two object arrays and have it delegate to
+, since logically those would be unrelated operations.
On Sep 18, 2015 00:36, "seberg" <notifications@github.com> wrote:

> Actually, I think it would likely make sense to define "object" arrays in
> the sense of "try to call the type specific ufunc -- if we know it -- on
> each element" + some fallback (unless they map to a python operator maybe).
> Just to annoy you, we might want to support numpy dtypes as well. Whether
> or not they may be cast, I am not sure ;). The ufuncs have slightly
> different behaviours sometimes I think, but yes, the only objections to it
> I can think of, is that is possibly a bit slow.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6320#issuecomment-141371053>.
>

seberg commented:
Too bad I refuse to read that discussion ;). Well, I think one reason to *not* have it the same would be to have elementwise vs. some other logic (such as concatenate)? In that sense, I think if we call `+` elementwise inside `np.add` (instead of the "elementwise `np.add`) that seems fine.
There would maybe be no way to "call elementwise add recursively" unless you specifically define the ufunc for your type.
The only thing I can think of that would break this type of logic is when you start wishing to propagate things like the `out` (or other) keyword argument. However, all of this kind of fancy tricks should be possibly if you implement `__numpy_ufunc__` for the container or probably more often have some specfic dtype functions for it even.

Which actually gets me philosophizing. Are maybe containers just dtype-scalars in some sense? Is there actually any difference between how a ufunc knows how to work on a single element and how `__numpy_ufunc__` would know how to work on a single "container"?

ahaldane committed:
Part 6 - fallback to np.ufunc

ahaldane commented:
Update. 

My previous thought wasn't quite right: We do need to take care of "long" specially. But now, if the incoming type is one that numpy knows how to deal with, just use the numpy ufunc. In the case of float and complex in particular, I think the numpy type is identical to the python type, so it works perfectly. This makes the implementations much less monstrous than in commit 5!

Also, I have gone back to a default implementation (instead of TypeError) for `max`, `min`, `fmax`, `fmin` and the `logical_` ufuncs, since it turns out a lot of numpy tests depend on the original fallback. In particular, arrays like `np.array([datetime.timedelta(5, 14)])` are actually object arrays and some tests want to take the max/min of such arrays.

Currently there is still one test that fails which I have commented out:

    >>> np.max(np.array(['a', 'b', 'c'], dtype=object))
    TypeError: ufunc 'maximum' did not contain a loop with signature matching types dtype('S1') dtype('S1') dtype('S1')

What happens is we end up calling `np.maximum('a', 'b')`, where numpy casts to `dtype('S1')`, and then numpy doesn't know how to `max` strings. Compare to:

    >>> np.max(np.array(['a', 'b', 'c']))
    TypeError: cannot perform reduce with flexible type

It is possible that unit test is wrong, or that reduce should work on strings.

ahaldane commented:
Oh, I intended for `numpy_knows` to return False for lists/ndarrays. That will have to be changed, later.

njsmith commented:
The issue with max-of-strings sounds serious to me -- thanks to pandas
there's a *lot* of code out there that expects object arrays full of
strings to work better than an actual string array.
.
Of course just fixing maximum to work on string arrays would also solve
part of the problem...
On Sep 18, 2015 12:46 PM, "ahaldane" <notifications@github.com> wrote:

> Oh, I intended for numpy_knows to return False for lists/ndarrays. That
> will have to be changed, later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6320#issuecomment-141549557>.
>

ahaldane committed:
Part 7

ahaldane commented:
Fixed the string issue. Now the code only treats (bool, int, long, float, complex) specially.

With this commit, the rules are:

1. If there is a special method corresponding to the ufunc, call it (this may change depending on what is decided for `__numpy_ufunc__`, as you are discussing).

2. For unary ufuncs, if the element is (a numpy scalar, `float`, or `complex`) simply call the numpy ufunc on the elements. If the element was (float or complex) convert back to the python type like `.item()` afterwards. For binary ufuncs, if either element is a numpy scalar or if both are (float or complex) use the numpy ufunc. If both elements are (float or complex), convert back to float/complex afterward.

3. If the first element has a method with the name of the ufunc, call that.

4. Call the fallback function from `_objectmath.py`. A few of these fallbacks take care of arbitrary types, but most only take care of python integers and fail otherwise.


Notes: 
 * Nested lists and ndarrays fall through to step 4, then usually fail.
 * Binary ufuncs called with an integer and a non-integer are a little tricky.. maybe the logic can be cleaned up, maybe not. 
 * Currently object arrays containing python integers are probably the slowest case since they have to go through the fallback. Maybe the integer fallback code should be moved to step 2.5 somehow.
 * @seberg, did you suggest I add a case to handle dtype objects? I know there was an idea that dtype instances would be the new scalars, but that's still hypothetical, right?

seberg commented:
I meant the regular numpy scalars (whatever that is) or even arrays. I.e. one could make a case for:
```
In [8]: a = np.empty(1, dtype=object)
In [9]: a[0] = np.float64(3)
In [10]: a
Out[10]: array([3.0], dtype=object)
In [11]: np.sqrt(a)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-11-4d1b1bf7082b> in <module>()
----> 1 np.sqrt(a)
AttributeError: 'numpy.float64' object has no attribute 'sqrt'
```
I am just happy to point out the mines here hoping you manage to figure it out :). Probably it is really good enough to support python types here. This does not (and probably cannot) get us 100%, so as long as it does not block the road ahead (and I think we got that now for sure), I am happy.

seberg commented:
Ah, should have read more careful, I guess you implemented the numpy scalar stuff exactly.

ahaldane committed:
Part 8

homu commented:
:umbrella: The latest upstream changes (presumably #6269) made this pull request unmergeable. Please resolve the merge conflicts.

ahaldane committed:
Part 9 (months later)

ahaldane committed:
fixups

mgeier committed:
TST: Tests for linalg.norm() using nested arrays

homu commented:
:umbrella: The latest upstream changes (presumably #7099) made this pull request unmergeable. Please resolve the merge conflicts.

ahaldane committed:
update comment

ahaldane committed:
fixups by rebase errors

ahaldane commented:
rebased. Still needs more work in the tests.

Now that the `__array_ufunc__` code is in (hooray!), I can try using it here to replace the named-method hack (step 3 in my last comment). I'll look into it later.

ahaldane committed:
isnan, isfinite, isinf, signbit should have bool out dtype

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2720,46 +2722,239 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
+            if (ret != NULL) {
+                arr = (PyArrayObject*)PyArray_FromScalar(ret, NULL);
+                Py_DECREF(ret);
+                ret = PyArray_GETITEM(arr, PyArray_DATA(arr));
```
What's going on here?

ahaldane replied:
For pure-python float and complex objects we call the ufunc on the object, which converts to a numpy type, eg float -> float64. Then we convert back to python types by calling `.item()`, but this method only exists on arrays so we need to convert the scalar to an array first, then call `PyArray_GETITEM`.

eric-wieser commented on numpy/core/_objectmath.py:
```diff
@@ -0,0 +1,262 @@
+    return bool(x and y)
+
+def logical_or(x, y):
+    return bool(x or y)
```
Can we leave the `bool` to a separate PR, since that's a behaviour change?

ahaldane replied:
Actually I want to go even further the other way. `logical_or` is documented to guarantee it returns a boolean array as output. 

I've already forced `isinf`, `isnan,` `isfinite` and `signbit` to return booleans, but it looks like I forgot to do so for the `logical_` ufuncs.

ahaldane replied:
But it's true maybe I can leave that for another PR... if I go all the way through with that idea I'd also have to make the comparison operators (`less` and so on) return bool arrays.

eric-wieser replied:
I'd be much keener to leave behaviour changes to a second PR, if possible - in particular, we might want a deprecation cycle, or even a way to explicitly request the old behaviour. But I guess we can review it all together, and split it up once everything else looks good.

ahaldane committed:
logical_* ufuncs should have bool out dtype

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2877,8 +2880,13 @@ OBJECT_@func@(char **args, npy_intp *dimensions, npy_intp *steps, void *data)
         }
          
+#if @out_bool@
+        *((npy_bool *)op1) = PyObject_IsTrue(ret);
```
This handles `-1` badly

eric-wieser replied:
We should just break the loop, as we do above

ahaldane replied:
good point, will fix

eric-wieser commented on numpy/core/_objectmath.py:
```diff
@@ -0,0 +1,262 @@
+        if isint(x):
+            return nfunc(float(x))
+        raise TypeError(_notimplemented_msg(name, type(x)))
+    return mathfunc
```
nitpick: set `mathfunc.__name__ == nfunc.__name__`

eric-wieser commented on numpy/core/code_generators/generate_umath.py:
```diff
@@ -360,15 +361,15 @@ def english_upper(s):
           TD(ints),
           TD(inexact, f='pow', astype={'e':'f'}),
-          TD(O, f='npy_ObjectPower'),
+          TD(O, f='PyNumber_Power'),
```
This doesn't look safe to me - we don't pass enough arguments to `PyNumber_Power` in the loop - `npy_ObjectPower` is there to add an extra `Py_None` argument

ahaldane replied:
yes, fixing

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2720,46 +2722,247 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
+           exp, expm1, log, log10, log1p, sqrt, ceil, trunc, fabs, floor,
+           arctan2, conjugate, square, reciprocal, _ones_like, sign, degrees,
+           rad2deg, radians, deg2rad, exp2, log2, cbrt, rint, spacing#
+   #out_bool = 1, 1, 1, 1, 1, 0*37#
```
Not a fan of this information being duplicated in python and C

eric-wieser commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2720,46 +2722,247 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
+    npy_cache_import("numpy.core._objectmath", "@func@", &fallback);
+    if (fallback == NULL) {
+        return;
+    }
```
Can we put these cache imports, and/or the function name, in `data` instead? Then we'd only need one loop function, rather than one for each ufunc (or perhaps two - one for boolean return values)

madphysicist commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -2720,46 +2722,247 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
+NPY_NO_EXPORT int
+_is_float_complex(PyObject *op)
+{
+    if (PyFloat_Check(op) || PyComplex_Check(op)) {
```
This won't work for objects that define `__float__` or `__complex__` methods.

eric-wieser commented on numpy/core/_objectmath.py:
```diff
@@ -0,0 +1,262 @@
+        if x == -1:
+            return -1
+        if x == 0:
+            return np.reciprocal(0).item()
```
This gives `RuntimeWarning: divide by zero encountered in reciprocal` - was that intended?

ahaldane replied:
I don't remember the intention anymore, but maybe. Doesn't the corresponding int32/int64 ufunc raise the warning too? Then the object ufunc probably should as well.

```python
np.reciprocal([1,0,1])                                                    
/usr/bin/ipython3:1: RuntimeWarning: divide by zero encountered in reciprocal
  #!/usr/bin/python3
/usr/bin/ipython3:1: RuntimeWarning: invalid value encountered in reciprocal
  #!/usr/bin/python3
array([                   1, -9223372036854775808,                    1])
```
